### ISSUE TITLE
Support `aten.max_pool2d_with_indices`

### ISSUE THREAD

**Issue by jdh8** [tenstorrent/tt-metal#12099]
As we find out in tenstorrent/pytorch2.0_ttnn#114, PyTorch converts `torch.nn.functional.max_pool2d` to `aten.max_pool2d_with_indices` even in indices are not requested.  For the sake of completeness and a simple structure, I suggest making a version of `max_pool2d` that simultaneously outputs indices.

**Comment by ayerofieiev-tt** [comment]
@jdh8 to clarify, this does not block compiler and you were able to lower to `ttnn.max_pool2d` for the case where indices are not used?

**Comment by jdh8** [comment]
I'm consulting @kevinwuTT for tech support on the workaround.
https://github.com/tenstorrent/pytorch2.0_ttnn/pull/114#discussion_r1738850006

I hope we can find a way to lower an expression involving `__getitem__`
```py
aten.max_pool2d_with_indices.default(x)[0]
```

**Comment by davorchap** [comment]
> I'm consulting @kevinwuTT for tech support on the workaround. [tenstorrent/pytorch2.0_ttnn#114 (comment)](https://github.com/tenstorrent/pytorch2.0_ttnn/pull/114#discussion_r1738850006)
> 
> I hope we can find a way to lower an expression involving `__getitem__`
> 
> ```python
> aten.max_pool2d_with_indices.default(x)[0]
> ```

Indicies won’t ready for a while (needed only for training ) so go with a workaround

**Comment by ayerofieiev-tt** [comment]
Removed from sprint, only needed for training. So far focusing more on inference

**Comment by mbezuljTT** [comment]
I've chatted with @pavlejosipovic on this

max_pool2d_with_indices ttnn support is required for OFT model that is priority for Shield team. 

timeline wise, we would need functional support by September, and perf optimization (if required) by end of October.

**Comment by mbezuljTT** [comment]
trace with shape we need 


Function | Input Shape | Output Shape | Attributes
-- | -- | -- | --
max_pool2d_with_indices | [(1, 1, 159, 159)] | [(1, 1, 159, 159), (1, 1, 159, 159)] | [[Tensor(shape=(1, 1, 159, 159), dtype=torch.float32), 3], {stride: 1, padding: 1, dilation: 1, ceil_mode: False, return_indices: True}]

**Comment by pavlejosipovic** [comment]
@wransom-TT lets investigate how hard would be to implement this, and what would be the ETA.

**Comment by mbezuljTT** [comment]
per offline chat, the size kernel 3x3

fwiw [reference code ](https://github.com/tom-roddick/oft/blob/cecedcfd858e613d5e500f42fd3469c7d289abb9/oft/data/encoder.py#L213)is implementing [non_maximum_suppression](https://docs.pytorch.org/vision/main/generated/torchvision.ops.nms.html) which is nowdays a torch op on it's own. We'll see if optimizing the op is on a critical path, if yes we might want to implement nms itself in the future.

**Comment by wransom-TT** [comment]
@cmaryanTT FYI

Hello, I have an update on the status of Pool2D with indices:

1. The trace mentioned above: `[[Tensor(shape=(1, 1, 159, 159), dtype=torch.float32), 3], {stride: 1, padding: 1, dilation: 1, ceil_mode: False, return_indices: True}]` is now working for `Bfloat16` on `Wormhole`.
2. Performance is quite slow (almost 30x slower than normal Pool2D) since we can only support reduction with indices on our hardware using TopK which leverages the SFPU instead of the FPU and is therefore very slow.
3. Beyond the performance there are a few other limitations currently:
- No blackhole support due to an LLK issue see this conversations: https://tenstorrent.slack.com/archives/C062KKRTHKR/p1755619923495069
- No support for Float32 (Pool2D does not support this even without indices)
- No support for large kernels with `kernel_h * kernel_w >= 32`
- No support for large tensors with `input_h * input_w >= 256^2`

Based on the trace I don't think the kernel size or input size limitations are a problem right now.  However I believe the lack of blackhole support, lack of Float32 support, and slow performance will be issues.  Here are some thoughts about those:
- This should work on blackhole, I am waiting to here back from the LLK team about this and will update when I have more information.
- Float32 support should be possible and won't be too difficult to implement but will take some additional development time and will likely result in even poorer kernel performance.
- LLK says that TopK does a lot of extra work which is not necessary for a reduction with indices so it may be possible to write a custom kernel which will be faster.

@mbezuljTT can you provide some insight on the criticalness of each of these problems?

**Comment by wransom-TT** [comment]
Closing this as basic MPWI now exists with planned generality and performance improvements tracked here:

https://github.com/tenstorrent/tt-metal/issues/27845
